File: 2016-05-02-265-easy-permutations-and-combinations-part-1/main.cpp
#include <iostream>
#include <numeric>
#include <string>
#include <vector>

std::string dec_to_fac(int n)
{
	int divisor = 1, rem;
	std::vector<int> vec;
	while (n != 0) {
		rem = n % divisor;
		n /= divisor++;
		vec.push_back(rem);
	}
	std::string ret;
	for (auto it = vec.rbegin(); it != vec.rend(); it++) {
		ret += std::to_string(*it);
	}
	return ret;
}

std::vector<int> make_permutation(std::vector<int> input, std::string fac)
{
	std::vector<int> ret;
	for (char c : fac) {
		int n = c - '0';
		ret.push_back(input.at(n));
		input.erase(input.begin() + n);
	}
	return ret;
}

int main(int argc, char *argv[])
{
	std::vector<int> vec1(6);
	std::iota(vec1.begin(), vec1.end(), 0);
	auto result1 = make_permutation(vec1, dec_to_fac(239)); // 240 - 1
	for (int n : result1) {
		std::cout << n << ' ';
	}
	std::cout << '\n';
	auto vec2 = std::vector<int>(7);
	std::iota(vec2.begin(), vec2.end(), 0);
	auto result2 = make_permutation(vec2, dec_to_fac(3239)); // 3240 - 1
	for (int n : result2) {
		std::cout << n << ' ';
	}
	std::cout << '\n';
	return 0;
}

File: 2016-05-02-265-easy-permutations-and-combinations-part-1/main.swift
import Foundation

func decimalToFactorial(fromDecimal dec: Int) -> String {
    var n = dec, divisor = 1, rem = 0
    var retArr = [String]()
    while n > 0 {
        rem = n % divisor
        retArr.append(String(rem))
        n /= divisor
        divisor += 1
    }
    return Array(retArr.reversed()).joined()
}

func makePermutation(fromNumbers numbers: [Int], fromFac fac: String) -> [Int] {
    var ret = [Int]()
    var mutInput = numbers
    for ch in fac {
        let n = Int(String(ch))!
        ret.append(mutInput[n])
        mutInput.remove(at: n)
    }
    return ret
}

let test1 = Array((0..<6))
let result1 = makePermutation(fromNumbers: test1, fromFac: decimalToFactorial(fromDecimal: 239))
print(result1)
let test2 = Array((0..<7))
let result2 = makePermutation(fromNumbers: test2, fromFac: decimalToFactorial(fromDecimal: 3239))
print(result2)
